### ISSUE TITLE
Add BenchmarkSpec

### ISSUE THREAD

**Issue by jdegoes** [zio/zio#3289]
A hypothetical benchmark spec would benchmark "tests" and report timing information, allowing configuration using aspects like `@@ warmup(50) @@ parallelism(5) @@ iterations(1000)`.

TODO: Spec this out more.

**Comment by adamgfraser** [comment]
I would like to work on this.

**Comment by jdegoes** [comment]
This will be super cool. It's important to minimize interference with benchmark specs, which means:

 — Only one of them running globally at any one time (no cross-spec parallelism)
 — Do a garbage collection prior to each run, to minimize GC inference
 — Minimizing per test overhead (`provide`, etc.); ideally we just repeat that test
 — Repeating the benchmark in a tight `flatMap` loop

A lot of the above annotations should be done "statically" on the initial annotation map.

**Comment by mijicd** [comment]
I'm not 100% sure this should be a regular spec. Perhaps it's better to have a separate "hierarchy" for benchmarks that could allow aspects (or combinators?) as the one mentioned above.

**Comment by adamgfraser** [comment]
Potentially. We had a different Spec for using suite level managed resources instead early draft. I tend to think we should try to do it as part of the same spec because we will get a lot of things “for free” and then if we run into issues we can consider falling back to that.

**Comment by mijicd** [comment]
True, but we need to make sure these tests are properly isolated. Benchmarks tend to be tedious and occasionally long running, so it might not be the best idea to allow composing them with regular ones.

**Comment by adamgfraser** [comment]
Yes. If we annotate the tests statically we could also have the test framework either fail or display a warning if benchmark tests are mixed with non-benchmark tests, or run all the benchmarks before or after running any other tests.

**Comment by hearnadam** [comment]
Not prioritized.
